### PR TITLE
Update logic of patch run to report / check status as it goes

### PR DESCRIPTION
**Pull request by ryanwe**
- Patching reports status and stops if the patch job is not active
- Handles restarts across crashes and reboots
- Handles dryruns
- Handles stopping on failure
- Supports running the agent as `ospatch` to facilitate testing.
- Does not *yet* handle retry attempts (that'll be another PR)
- I did a lot of manual testing but this code doesn't lend itself well to unit testing. I need to think more about how to do that in a subsequent PR.

### CODE REVIEW

File: cli_tools/google-osconfig-agent/patch/patch_run.go [hunk 1]
```diff
@@ -16,38 +16,68 @@ package patch
...
+	go watcher(ctx, savedPatchJobName)
+}
+
+// RunPatchAgent runs patching as a single blocking agent.
+func RunPatchAgent(ctx context.Context) {
+	savedPatchJobName := checkSavedState(ctx)
+	watcher(ctx, savedPatchJobName)
+}
+
+// Load current patch state off disk and return its name.
```
Comment (adjackura) [inline cli_tools/google-osconfig-agent/patch/patch_run.go]: This also runs a patch if its in the middle of one.

File: cli_tools/google-osconfig-agent/patch/patch_run.go [hunk 2]
```diff
@@ -16,38 +16,68 @@ package patch
...
 	"context"
+	"fmt"
 	"os"
 	"time"
 
-	"github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/google-osconfig-agent/_internal/gapi-cloud-osconfig-go/cloud.google.com/go/osconfig/apiv1alpha1"
-	osconfigpb "github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/google-osconfig-agent/_internal/gapi-cloud-osconfig-go/google.golang.org/genproto/googleapis/cloud/osconfig/v1alpha1"
+	"cloud.google.com/go/compute/metadata"
+	osconfig "github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/google-osconfig-agent/_internal/gapi-cloud-osconfig-go/cloud.google.com/go/osconfig/apiv1alpha1"
+	api "github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/google-osconfig-agent/_internal/gapi-cloud-osconfig-go/google.golang.org/genproto/googleapis/cloud/osconfig/v1alpha1"
```
Comment (adjackura) [inline cli_tools/google-osconfig-agent/patch/patch_run.go]: osconfigpb is more correct

File: cli_tools/google-osconfig-agent/patch/patch_run.go [hunk 3]
```diff
@@ -16,38 +16,68 @@ package patch
...
+	completeSuccess
+	completeFailed
+	completeJobCompleted
+)
+
 // Init starts the patch system.
 func Init(ctx context.Context) {
-	savedPatchName := ""
-	// Load current patch state off disk.
+	savedPatchJobName := checkSavedState(ctx)
```
Comment (adjackura) [inline cli_tools/google-osconfig-agent/patch/patch_run.go]: go RunPatchAgent ?

File: cli_tools/google-osconfig-agent/patch/patch_run.go [hunk 4]
```diff
@@ -77,89 +109,251 @@ func (j *patchJob) UnmarshalJSON(b []byte) error {
...
+}
+
+func ackPatch(ctx context.Context, patchJobName string) {
+	currentPatchJob, err := loadState(state)
+	if err != nil {
+		logger.Errorf("Unable to load state to ack notification. Error: %v", err)
+		return
+	}
+
+	// Notify the server if we haven't yet
```
Comment (adjackura) [inline cli_tools/google-osconfig-agent/patch/patch_run.go]: Change this logic to not enqueue if the patch is already acked

Also add a TODO to store the already acked patch runs in memory so we don't have to read the state file.

File: cli_tools/google-osconfig-agent/patch/patch_run.go [hunk 5]
```diff
@@ -16,38 +16,68 @@ package patch
...
+const (
+	identityTokenPath = "instance/service-accounts/default/identity?audience=osconfig.googleapis.com&format=full"
+)
+
+type patchStep int
+
+// These patch steps are ordered to allow us to skip steps that have already been completed. Progress should
+// always move forward, even if we need to retry a step.
+const (
+	unknown patchStep = iota
```
Comment (adjackura) [inline cli_tools/google-osconfig-agent/patch/patch_run.go]: Reduce these to just the ones we act on currently, then add TODO to refactor into steps and step types

File: cli_tools/google-osconfig-agent/patch/patch_run.go [hunk 6]
```diff
@@ -16,38 +16,68 @@ package patch
...
 	"context"
+	"fmt"
 	"os"
 	"time"
 
-	"github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/google-osconfig-agent/_internal/gapi-cloud-osconfig-go/cloud.google.com/go/osconfig/apiv1alpha1"
-	osconfigpb "github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/google-osconfig-agent/_internal/gapi-cloud-osconfig-go/google.golang.org/genproto/googleapis/cloud/osconfig/v1alpha1"
+	"cloud.google.com/go/compute/metadata"
+	osconfig "github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/google-osconfig-agent/_internal/gapi-cloud-osconfig-go/cloud.google.com/go/osconfig/apiv1alpha1"
+	osconfigpg "github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/google-osconfig-agent/_internal/gapi-cloud-osconfig-go/google.golang.org/genproto/googleapis/cloud/osconfig/v1alpha1"
```
Comment (adjackura) [inline cli_tools/google-osconfig-agent/patch/patch_run.go]: osconfigpb

Comment (ryanwe) [reply cli_tools/google-osconfig-agent/patch/patch_run.go]: Thanks for catching that! Fixed.

**Pull request merged**